README.md
# Cheripic

[![Gem Version](https://badge.fury.io/rb/cheripic.svg)](https://badge.fury.io/rb/cheripic)
[![Build Status](https://travis-ci.org/shyamrallapalli/cheripic.svg?branch=master)](https://travis-ci.org/shyamrallapalli/cheripic)
[![Coverage Status](https://coveralls.io/repos/github/shyamrallapalli/cheripic/badge.svg?branch=master)](https://coveralls.io/github/shyamrallapalli/cheripic?branch=master)
[![Code Climate](https://codeclimate.com/github/shyamrallapalli/cheripic/badges/gpa.svg)](https://codeclimate.com/github/shyamrallapalli/cheripic)



Computing Homozygosity Enriched Regions In genomes to Prioritize Identification of Candidate variants (CHERIPIC), 
is a ruby tools to pick causative mutation from bulks segregant sequencing.     
        
Currently this gem is still in development and nearing complete working package.
And software only works with pileup as input files, use of bam and vcf files will be implemented in future
        
        
## Installation

Cheripic is available both as a command line tool and as a gem.     
Binaries are available for Linux 64bit and OSX.      
Best way to use Cheripic is to download appropriate binary arhcive      
unpack (`tar -xzf`) and add the unpacked directory to your `PATH`       

Latest binaries are available to [download here](https://github.com/shyamrallapalli/cheripic/releases/latest)       


To install gem and use the gem in your development     
Add this line to your application's Gemfile:

```ruby
gem 'cheripic'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install cheripic

## Usage

Running `cheripic` without any input at command line interface shows following help options

```

Cheripic v1.2.6
Authors: Shyam Rallapalli and Dan MacLean

Description: Candidate mutation and closely linked marker selection for non reference genomes
Uses bulk segregant data from non-reference sequence genomes

Inputs:
1. Needs a reference fasta file of asssembly use for variant analysis
2. Pileup/Bam files for mutant (phenotype of interest) bulks and background (wildtype phenotype) bulks
3. If providing bam files, you have to include vcf files for the respective bulks
4. If polyploid species, include pileup/bam files from one or both parents

USAGE:
cheripic <options>

OPTIONS:
  -f, --assembly=<s>               Assembly file in FASTA format
  -F, --input-format=<s>           bulk and parent alignment file format types - set either pileup or bam or vcf (default: pileup)
  -a, --mut-bulk=<s>               Pileup or sorted BAM file alignments from mutant/trait of interest bulk 1
  --mut-bulk-vcf=<s>               vcf file for variants from mutant/trait of interest bulk 1 (default: )
  -b, --bg-bulk=<s>                Pileup or sorted BAM file alignments from background/wildtype bulk 2
  --bg-bulk-vcf=<s>                vcf file for variants from background/wildtype bulk 2 (default: )
  --output=<s>                     custom name tag to include in the output file name (default: cheripic_results)
  --loglevel=<s>                   Choose any one of "info / warn / debug" level for logs generated (default: debug)
  --hmes-adjust=<f>                factor added to snp count of each contig to adjust for hme score calculations (default: 0.5)
  --htlow=<f>                      lower level for categorizing heterozygosity (default: 0.2)
  --hthigh=<f>                     high level for categorizing heterozygosity (default: 0.9)
  --mindepth=<i>                   minimum read depth at a position to consider for variant calls (default: 6)
  --max-d-multiple=<i>             multiplication factor for average coverage to calculate maximum read coverage
                                   if set zero no calculation will be made from bam file.
                                   setting this value will override user set max depth (Default: 5)
  --maxdepth=<i>                   maximum read depth at a position to consider for variant calls
                                   if set to zero no user max depth will be used (default: 0)
  --min-non-ref-count=<i>          minimum read depth supporting non reference base at each position (default: 3)
  --min-indel-count-support=<i>    minimum read depth supporting an indel at each position (default: 3)
  --ambiguous-ref-bases=<s>        including variant at completely ambiguous bases in the reference (default: false)
  -q, --mapping-quality=<i>        minimum mapping quality of read covering the position (default: 20)
  -Q, --base-quality=<i>           minimum base quality of bases covering the position (default: 15)
  --noise=<f>                      praportion of reads for a variant to conisder as noise (default: 0.1)
  --cross-type=<s>                 type of cross used to generated mapping population - back or out (default: back)
  --use-all-contigs=<s>            option to select all contigs or only contigs containing variants for analysis (default: false)
  --include-low-hmes=<s>           option to include or discard variants from contigs with
                                   low hme-score or bfr score to list in the final output (default: false)
  --polyploidy=<s>                 Set if the data input is from polyploids (default: false)
  -p, --mut-parent=<s>             Pileup or sorted BAM file alignments from mutant/trait of interest parent (default: )
  -r, --bg-parent=<s>              Pileup or sorted BAM file alignments from background/wildtype parent (default: )
  -R, --repeats-file=<s>           repeat masker output file for the assembly  (default: )
  --bfr-adjust=<f>                 factor added to hemi snp frequency of each parent to adjust for bfr calculations (default: 0.05)
  --sel-seq-len=<i>                sequence length to print from either side of selected variants (default: 50)
  --examples                       shows some example commands with explanation

```
        
        
        
Example Commands


```
EXAMPLE COMMANDS:
  1. cheripic -f assembly.fa -a mutbulk.pileup -b bgbulk.pileup --output=cheripic_output
  2. cheripic --assembly assembly.fa --mut-bulk mutbulk.pileup --bg-bulk bgbulk.pileup 
        --mut-parent mutparent.pileup --bg-parent bgparent.pileup --polyploidy true --output cheripic_results
  3. cheripic --assembly assembly.fa --mut-bulk mutbulk.pileup --bg-bulk bgbulk.pileup 
        --mut-parent mutparent.pileup --bg-parent bgparent.pileup --polyploidy true 
        --use-all-contigs true --include-low-hmes true --output cheripic_results

```


By default contigs with out a variant and thos contigs with lower scores are discarded. 
      so use options `--no-only-frag-with-vars` and `--no-filter-out-low-hmes` to disable them 


## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/shyamrallapalli/cheripic. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](contributor-covenant.org) code of conduct.


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



ChangeLog.md
### Change Log

All significant changes to this project at each release are documented in this file.


#### Future changes to include
    
    1. option to take multiple background pileup files

#### [1.2.6] - 2016-10-26

    1. option to run only using with vcf files as bulk inputs to increase speed of analysis for larger genomes

#### [1.2.5] - 2016-10-17

    1. Updated methods take bam file (along with a vcf file) or pileup file as inputs of bulks
    2. Replaced output directory with output file name tag, since we only write to one file

#### [1.2.0] - 2016-08-11

    1. fixed calculation of heterzygosity for background bulks
    2. changed command line boolean option to be set using only true or false
    3. included command line option to set length of sequnce to retireve on either side of each variant
     
    
#### [1.1.0] - 2016-07-26

    first release of the binaries for Linux 64 bit and OSX 64bit


lib/cheripic/cmd.rb
#!/usr/bin/env ruby

module Cheripic

  # A command line option and processing object to handle input options
  #
  # @!attribute [rw] options
  #   @return [Hash] a hash of trollop option names as keys and user or default setting as values
  class Cmd

    require 'trollop'
    require 'pathname'
    require 'ostruct'

    attr_accessor :options

    # creates a Cmd object using input string entry
    # @param args [String]
    def initialize(args)
      @options = parse_arguments(args)
      check_arguments
    end

    # method to check input command string and run appropriate
    # method of the object (help or examples or parsing arguments)
    # @param args [String]
    def parse_arguments(args)
      Trollop::with_standard_exception_handling argument_parser do
        if args.empty? || args.include?('-h') || args.include?('--help')
          raise Trollop::HelpNeeded
        elsif args.include?('--examples')
          print_examples
        end
        argument_parser.parse args
      end
    end

    # trollop argument_parser for input args string and
    # @return [Hash] a hash of trollop option names as keys and user or default setting as values
    def argument_parser
      cmds = self
      Trollop::Parser.new do
        version Cheripic::VERSION
        banner cmds.help_message
        opt :assembly, 'Assembly file in FASTA format',
            :short => '-f',
            :type => String
        opt :input_format, 'bulk and parent alignment file format types - set either pileup or bam or vcf',
            :short => '-F',
            :type => String,
            :default => 'pileup'
        opt :mut_bulk, 'Pileup or sorted BAM file alignments from mutant/trait of interest bulk 1',
            :short => '-a',
            :type => String
        opt :mut_bulk_vcf, 'vcf file for variants from mutant/trait of interest bulk 1',
            :type => String,
            :default => ''
        opt :bg_bulk, 'Pileup or sorted BAM file alignments from background/wildtype bulk 2',
            :short => '-b',
            :type => String
        opt :bg_bulk_vcf, 'vcf file for variants from background/wildtype bulk 2',
            :type => String,
            :default => ''
        opt :output, 'custom name tag to include in the output file name',
            :default => 'cheripic_results'
        opt :loglevel, 'Choose any one of "info / warn / debug" level for logs generated',
            :default => 'info'
        opt :hmes_adjust, 'factor added to snp count of each contig to adjust for hme score calculations',
            :type => Float,
            :default => 0.5
        opt :htlow, 'lower level for categorizing heterozygosity',
            :type => Float,
            :default => 0.2
        opt :hthigh, 'high level for categorizing heterozygosity',
            :type => Float,
            :default => 0.9
        opt :mindepth, 'minimum read depth at a position to consider for variant calls',
            :type => Integer,
            :default => 6
        opt :max_d_multiple, "multiplication factor for average coverage to calculate maximum read coverage
if set zero no calculation will be made from bam file.\nsetting this value will override user set max depth",
            :type => Integer,
            :default => 5
        opt :maxdepth, 'maximum read depth at a position to consider for variant calls
if set to zero no user max depth will be used',
            :type => Integer,
            :default => 0
        opt :min_non_ref_count, 'minimum read depth supporting non reference base at each position',
            :type => Integer,
            :default => 3
        opt :min_indel_count_support, 'minimum read depth supporting an indel at each position',
            :type => Integer,
            :default => 3
        opt :ambiguous_ref_bases, 'including variant at completely ambiguous bases in the reference',
            :type => String,
            :default => 'false'
        opt :mapping_quality, 'minimum mapping quality of read covering the position',
            :short => '-q',
            :type => Integer,
            :default => 20
        opt :base_quality, 'minimum base quality of bases covering the position',
            :short => '-Q',
            :type => Integer,
            :default => 15
        opt :noise, 'praportion of reads for a variant to conisder as noise',
            :type => Float,
            :default => 0.1
        opt :cross_type, 'type of cross used to generated mapping population - back or out',
            :type => String,
            :default => 'back'
        opt :use_all_contigs, 'option to select all contigs or only contigs containing variants for analysis',
            :type => String,
            :default => 'false'
        opt :include_low_hmes, 'option to include or discard variants from contigs with
low hme-score or bfr score to list in the final output',
            :type => String,
            :default => 'false'
        opt :polyploidy, 'Set if the data input is from polyploids',
            :type => String,
            :default => 'false'
        opt :mut_parent, 'Pileup or sorted BAM file alignments from mutant/trait of interest parent',
            :short => '-p',
            :type => String,
            :default => ''
        opt :bg_parent, 'Pileup or sorted BAM file alignments from background/wildtype parent',
            :short => '-r',
            :type => String,
            :default => ''
        opt :repeats_file, 'repeat masker output file for the assembly ',
            :short => '-R',
            :type => String,
            :default => ''
        opt :bfr_adjust, 'factor added to hemi snp frequency of each parent to adjust for bfr calculations',
            :type => Float,
            :default => 0.05
        opt :sel_seq_len, 'sequence length to print from either side of selected variants',
            :type => Integer,
            :default => 50
        opt :examples, 'shows some example commands with explanation'
      end
    end

    # help message to display from command line
    def help_message
      msg = <<-EOS

      Cheripic v#{Cheripic::VERSION.dup}
      Authors: Shyam Rallapalli and Dan MacLean

      Description: Candidate mutation and closely linked marker selection for non reference genomes
      Uses bulk segregant data from non-reference sequence genomes

      Inputs:
      1. Needs a reference fasta file of asssembly use for variant analysis
      2. Pileup/Bam files for mutant (phenotype of interest) bulks and background (wildtype phenotype) bulks
      3. If providing bam files, you have to include vcf files for the respective bulks
      4. If polyploid species, include pileup/bam files from one or both parents

      USAGE:
      cheripic <options>

      OPTIONS:

      EOS
      msg.split("\n").map{ |line| line.lstrip }.join("\n")
    end

    # examples to display from command line
    def print_examples
      msg = <<-EOS

      Cheripic v#{Cheripic::VERSION.dup}
      Authors: Shyam Rallapalli and Dan MacLean

      EXAMPLE COMMANDS:
        1. cheripic -f assembly.fa -a mutbulk.pileup -b bgbulk.pileup --output=cheripic_output
        2. cheripic --assembly assembly.fa --mut-bulk mutbulk.pileup --bg-bulk bgbulk.pileup
              --mut-parent mutparent.pileup --bg-parent bgparent.pileup --polyploidy true --output cheripic_results
        3. cheripic --assembly assembly.fa --mut-bulk mutbulk.pileup --bg-bulk bgbulk.pileup
              --mut-parent mutparent.pileup --bg-parent bgparent.pileup --polyploidy true
              --no-only-frag-with-vars --no-filter-out-low-hmes --output cheripic_results
        4. cheripic -h or cheripic --help
        5. cheripic -v or cheripic --version

      EOS
      puts msg.split("\n").map{ |line| line.lstrip }.join("\n")
      exit(0)
    end

    # calls other methods to check if command line inputs are valid
    def check_arguments
      convert_boolean_strings
      check_output
      check_log_level
      check_input_entry
      check_input_types
    end

    # convert true or false options to boolean
    def convert_boolean_strings
      %i{ambiguous_ref_bases use_all_contigs include_low_hmes polyploidy}.each do | symbol |
        if @options.key?(symbol)
          @options[symbol] = @options[symbol] == 'false' ? false : true
        end
      end
    end

    # set file given option to false if input is nil or None or ''
    def check_input_entry
      %i{assembly mut_bulk bg_bulk mut_bulk_vcf bg_bulk_vcf mut_parent bg_parent repeats_file}.each do | symbol |
        if @options.key?(symbol)
          if @options[symbol] == 'None'
            param = (symbol.to_s + '_given').to_sym
            @options[symbol] = ''
            @options.delete(param)
          end
        end
      end
    end

    # checks input files based on bulk file type
    def check_input_types
      inputfiles = {}
      inputfiles[:required] = %i{assembly mut_bulk}
      inputfiles[:optional] = %i{bg_bulk}
      if @options[:input_format] == 'bam'
        inputfiles[:required] << %i{mut_bulk_vcf}
        inputfiles[:optional] << %i{bg_bulk_vcf}
      end
      if @options[:polyploidy]
        inputfiles[:either] = %i{mut_parent bg_parent}
      end
      check_input_files(inputfiles)
    end

    # checks if input files are valid
    def check_input_files(inputfiles)
      inputfiles.each_key do | type |
        inputfiles[type].flatten!
        check = 0
        inputfiles[type].each do | symbol |
          if @options[symbol] == nil or @options[symbol] == ''
            if type == :required
              raise CheripicArgError.new "Options #{inputfiles}, all must be specified. Try --help for further help."
            end
          else
            file = @options[symbol]
            if symbol == :bg_bulk or symbol == :bg_bulk_vcf
              if file.include? ','
                @options[symbol] = []
                file.split(',').each do | infile |
                  @options[symbol] << File.expand_path(infile)
                  file_exist?(symbol, infile)
                end
              end
            else
              @options[symbol] = File.expand_path(file)
              file_exist?(symbol, file)
            end
            check = 1
          end
        end
        if type == :either and check == 0
          raise CheripicArgError.new "One of the options #{inputfiles}, must be specified. " +
                                       'Try --help for further help.'
        end
      end
    end

    def file_exist?(symbol, file)
      # checks if a given file exists
      unless File.exist?(file)
        raise CheripicIOError.new "#{symbol} file, #{file} does not exist!"
      end
    end

    # checks if files with output tag name already exists
    def check_output
      if (@options[:output].split('') & %w{# / : * ? ' < > | & $ ,}).any?
        raise CheripicArgError.new 'please choose a name tag that contains ' +
                                       'alphanumeric characters, hyphen(-) and underscore(_) only'
      end
      @options[:hmes_frags] = "#{@options[:output]}_selected_hme_variants.txt"
      @options[:bfr_frags] = "#{@options[:output]}_selected_bfr_variants.txt"
      [@options[:hmes_frags], @options[:bfr_frags]].each do | file |
        if File.exist?(file)
          raise CheripicArgError.new "'#{file}' file exists " +
                                         'please choose a different name tag to be included in the output file name'
        end
      end
    end

    # checks and sets logger level
    def check_log_level
      unless %w(error info warn debug).include?(@options[:loglevel])
        raise CheripicArgError.new "Loglevel #{@options[:loglevel]} is not valid. " +
                                       'It must be one of: error, info, warn, debug.'
      end
      logger.level = Yell::Level.new @options[:loglevel].to_sym
    end

    # Initializes an Implementer object using input options
    # and calls run method of the Implementer to start the pipeline
    # A hash of trollop option names as keys and user or default
    # setting as values is passed to Implementer object
    def run
      @options[:hmes_frags] = File.expand_path @options[:hmes_frags]
      @options[:bfr_frags] = File.expand_path @options[:bfr_frags]
      analysis = Implementer.new(@options)
      analysis.run
    end

  end # Cmd

end # Cheripic
